### PR TITLE
Provide completion on dwim-directories (#1820)

### PR DESCRIPTION
**Pull request by thierryvolpiatto**
* helm-files.el (helm-dwim-target): New.
(helm-dwim-target-directory): Use it.

### CODE REVIEW

File: helm-files.el [hunk 1]
```diff
@@ -627,22 +627,48 @@ Should not be used among other sources.")
...
+  :type '(radio :tag "Define default target directory for file actions."
+          (const :tag "Directory belonging to next window" next-window)
+          (const :tag "Completion on directories belonging to each window" completion)
+          (const :tag "Use initial directory or `default-directory'" nil)))
+
 (defun helm-dwim-target-directory ()
   "Return value of `default-directory' of buffer in other window.
 If there is only one window return the value of currently visited directory
 if found in `helm-ff-history' or fallback to `default-directory'
 of current buffer."
```
Comment (Ambrevar) [inline helm-files.el]: You should mention `helm-dwim-target` here.

**Comment by Ambrevar** [comment]
Looks good to me:
- `'completion` is awesome.
- The `nil` behaviour is possibly the most intuitive. Nice to have that available too.

Some documentation in `C-h m` and this is perfect.

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

> Looks good to me:
>
>   * 'completion is awesome.
>   * The nil behaviour is possibly the most intuitive. Nice to have that available too.

I hesitate making this the default WDYT ?

> Some documentation in C-h m and this is perfect.

Done, would be better though, let me know.
Also I had a doubt for prompt name in completion which is "Target
directory", if you think at a more meaningful name, let me know too.
 
Thanks.

-- 
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by Ambrevar** [comment]
> I hesitate making this the default WDYT ?

`'completion' is awesome but spawns an additional Helm buffer, so that's a total of 3 Helm buffers for the copy/move actions.  It can be confusing to beginners.
I think `nil` is more straightforward.  I would make it the default.

> Also I had a doubt for prompt name in completion which is "Target
> directory", if you think at a more meaningful name, let me know too.

Indeed, I don't like the name either, a newcomer will probably wonder what's the difference between the target and the file name.
I don't think we should come up with a new _definition_, we could rather provide an explicit sentence, such as:
- "Browse target starting from:", or "Select target starting from:"
- "Switch to folder before selecting target"
- "Base folder for target selection"

**Comment by Ambrevar** [comment]
Wait, looks like there is some unexpected side effect: with 2 windows open, if I ediff from helm-find-files (`C-c =`) I get prompted for a target directory.  I think this should not happen.

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

>     I hesitate making this the default WDYT ?
>
> 'completion' is awesome but spawns an additional Helm buffer, so
> that's a total of 3 Helm buffers for the copy/move actions. It can be
> confusing to beginners.

Agreed.

> I think `nil` is more straightforward. I would make it the default.

Done.

>     Also I had a doubt for prompt name in completion which is "Target
>     directory", if you think at a more meaningful name, let me know too.
>
> Indeed, I don't like the name either, a newcomer will probably wonder
> what's the difference between the target and the file name.  I don't
> think we should come up with a new definition, we could rather provide
> an explicit sentence, such as:
>
>   * "Browse target starting from:",

I took this one.

Thanks for the feedback, I will merge now.

--
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

> Wait, looks like there is some unexpected side effect: with 2 windows
> open, if I ediff from helm-find-files (C-c =) I get prompted for a
> target directory. I think this should not happen.

Indeed, I forget to fix it with new behavior, should work same as before
now.

Thanks.

--
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Pull request merged**